node/core/dispute-coordinator/src/real/initialized.rs

```diff
@@ -277,7 +293,36 @@ impl Initialized {
 				},
 				Ok(SessionWindowUpdate::Unchanged) => {},
 			};
-			self.scrape_on_chain_votes(ctx, overlay_db, new_leaf, now).await?;
+
+			// Scrape the head if above rolling session update went well.
+			if self.error.is_none() {
+				self.scrape_on_chain_votes(ctx, overlay_db, new_leaf.hash, now).await?;
+			}
+
+			// Try to scrape any blocks for which we could not get the current session or did not receive an
+			// active leaves update.
+			let target_block = new_leaf.number.saturating_sub(MAX_BATCH_SCRAPE_ANCESTORS);
+			let ancestors = OrderingProvider::get_block_ancestors(
+				ctx.sender(),
+				new_leaf.hash,
+				new_leaf.number,
+				target_block,
+				&mut self.last_scraped_blocks,
+			)
+			.await
+			.unwrap_or_else(|err| {
+				tracing::debug!(
+					target: LOG_TARGET,
+					"Skipping leaf ancestors scraping due to error: {}",
+					err
+				);
+				Vec::new()
+			});
+
+			// Maybe run these in parallel ?
```

Comment:
It's fine to do it serially, there is a limit on concurrent wasm instances iirc.